### PR TITLE
[MRG] Bayesian Optimization Tuner

### PR DESCRIPTION
**Pull request by jhfjhfj1**
Bayesian Optimization Tuner
Work in progress.

Pull request for convenience of sharing the code.
Not ready for review

### CODE REVIEW

**Comment by fchollet** [comment]
You'll need to rebase the PR; currently it contains a lot of changes that were made in `master`.

Also, please target the branch `master` rather than `refactor`.

File: kerastuner/tuners/bayesian.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+from ..engine import tuner as tuner_module
+from ..engine import oracle as oracle_module
+from ..engine import hyperparameters as hp_module
+
+
+class BayesianOptimizationOracle(oracle_module.Oracle):
+    """Bayesian optimization oracle.
+
+    It uses Bayesian optimization with a underlying Gaussian process model.
+    The acquisition function used is upper confidence bound (UCB).
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: Is there is a paper reference for this, it would be good to put the link in the docstring

File: kerastuner/tuners/bayesian.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        alpha: Float. Value added to the diagonal of the kernel matrix
+            during fitting.
+        beta: Float. The balancing factor of exploration and exploitation.
+            The larger it is, the more explorative it is.
+        seed: Int. Random seed.
+
+    """
+
+    def __init__(self,
+                 init_samples=2,
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: It is not immediately clear that `init` should be a number, or that "samples" refers to sets of hp values in the search space. What could be a better name? `num_initial_configurations`? `num_initial_points`?

File: kerastuner/tuners/bayesian.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+class BayesianOptimizationOracle(oracle_module.Oracle):
+    """Bayesian optimization oracle.
+
+    It uses Bayesian optimization with a underlying Gaussian process model.
+    The acquisition function used is upper confidence bound (UCB).
+
+    Attributes:
+        init_samples: Int. The number of randomly generated samples as initial
+            training data for Bayesian optimization.
+        alpha: Float. Value added to the diagonal of the kernel matrix
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: The docstring needs information about what the user is supposed to do with the argument (and if the user isn't supposed to touch it, it shouldn't be an argument). For instance, below we have "The larger it is, the more explorative it is.", which is useful. Docstrings should be actionable.

File: kerastuner/tuners/bayesian.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    def populate_space(self, trial_id, space):
+        self.update_space(space)
+        # Generate enough samples before training Gaussian process.
+        if self._num_trials < self.init_samples or len(self._score) < 2:
+            self._num_trials += 1
+            values = self._new_trial()
+            self._values[trial_id] = values
+            return {'status': 'RUN', 'values': values}
+
+        self._training_data()
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: Would a more explicit approach be more readable here? e.g. `y, x = self._get_training_data(values); self.gpr.fit(x, y)`

Currently it's not clear from reading the flow what `self._training_data()` is supposed to do.

File: kerastuner/tuners/bayesian.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+            (or callable that takes hyperparameters
+            and returns a Model isntance).
+        objective: String. Name of model metric to minimize
+            or maximize, e.g. "val_accuracy".
+        max_trials: Int. Total number of trials
+            (model configurations) to test at most.
+            Note that the oracle may interrupt the search
+            before `max_trial` models have been tested.
+        init_samples: Int. The number of randomly generated samples as initial
+            training data for Bayesian optmization.
```
Comment (floscha) [inline kerastuner/tuners/bayesian.py]: Typo. Should be `optimization`.

File: kerastuner/tuners/bayesian.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+    def __init__(self,
+                 hypermodel,
+                 objective,
+                 max_trials,
+                 init_samples=2,
+                 seed=None,
+                 **kwargs):
+        oracle = BayesianOptimizationOracle(init_samples=init_samples,
+
```
Comment (floscha) [inline kerastuner/tuners/bayesian.py]: This line needs to be removed.

File: kerastuner/tuners/bayesian.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        return min_x.reshape(-1, 1)
+
+
+class BayesianOptimization(tuner_module.Tuner):
+    """BayesianOptimization tuning with Gaussian process.
+
+    Args:
+        hypermodel: Instance of HyperModel class
+            (or callable that takes hyperparameters
+            and returns a Model isntance).
```
Comment (floscha) [inline kerastuner/tuners/bayesian.py]: Typo. Should be `instance`.

File: kerastuner/tuners/bayesian.py (hunk 2 above)
Comment (jhfjhfj1) [reply kerastuner/tuners/bayesian.py]: I think `num_initial_points` is better.

File: kerastuner/tuners/bayesian.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+            for index, value in enumerate(vector):
+                if not isinstance(self.space[index], hp_module.Fixed):
+                    no_fixed_vector.append(value)
+            x.append(no_fixed_vector)
+            y.append(score)
+        x = np.array(x)
+        y = np.array(y) / np.max(y)
+        return x, y
+
+    def _convert(self, vector):
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: This method should be named more explicitly. Avoid overly generic names.

File: kerastuner/tuners/bayesian.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        except UserWarning:
+            mu = np.array([1.0] * len(x), dtype=np.float64)
+            sigma = np.array([0.1] * len(x), dtype=np.float64)
+        return mu - self.beta * sigma
+
+    @property
+    def _dim(self):
+        return len([hp for hp in self.space if not isinstance(hp, hp_module.Fixed)])
+
+    def _generate_vector(self, ):
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: remove `, `

File: kerastuner/tuners/bayesian.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+    def populate_space(self, trial_id, space):
+        self.update_space(space)
+        # Generate enough samples before training Gaussian process.
+        if self._num_trials < self.num_initial_points or len(self._score) < 2:
+            self._num_trials += 1
+            values = self._new_trial()
+            self._values[trial_id] = values
+            return {'status': 'RUN', 'values': values}
+
+        # Update Gaussian process with existing samples
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: The lines below up to `values = ...` should probably be abstracted in a separate method e.g. `self._sample_values()`

File: kerastuner/tuners/bayesian.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+        x, y = self._get_training_data()
+        self.gpr.fit(x, y)
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: Is it possible to employ incremental fitting instead of batch fitting (retraining on all the available data every time we need to sample), and if so, shouldn't we incrementally fit in `result` rather than in `populate`?

File: kerastuner/tuners/bayesian.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        for index, hp in enumerate(self.space):
+            if hp.name == name:
+                return index
+        return None
+
+    def _upper_confidence_bound(self, x):
+        x = x.reshape(-1, self._dim)
+        try:
+            mu, sigma = self.gpr.predict(x, return_std=True)
+        except UserWarning:
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: Warnings are not exceptions (at least not by default). Is this tested?

Perhaps there is a simpler way to detect when default values should be used? E.g. if no training data is available.

File: kerastuner/tuners/bayesian.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+    """
+
+    def __init__(self,
+                 hypermodel,
+                 objective,
+                 max_trials,
+                 init_samples=2,
+                 seed=None,
+                 **kwargs):
+        oracle = BayesianOptimizationOracle(num_initial_points=init_samples,
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: This argument should have the same name across the tuner and the oracle (init_samples)

File: kerastuner/tuners/bayesian.py (hunk 11 above)
Comment (jhfjhfj1) [reply kerastuner/tuners/bayesian.py]: I will move it to result. but the incremental fitting with various lengths is not supported by the sklearn gaussian process regressor. We need to implement our own to support it.

File: kerastuner/tuners/bayesian.py (hunk 12 above)
Comment (jhfjhfj1) [reply kerastuner/tuners/bayesian.py]: It was caused by large values of y and it is fixed by normalizing y. I will remove the except.

File: kerastuner/tuners/bayesian.py [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+        self._values[trial_id] = values
+        return {'status': 'RUN', 'values': values}
+
+    def report_status(self, trial_id, status):
+        # TODO
+        raise NotImplementedError
+
+    def save(self, fname):
+        state = {
+            'init_samples': self.num_initial_points,
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: This is inconsistent, it should be called `num_initial_points` in the config as well
